flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java

```diff
@@ -605,6 +624,289 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
+                });
+        taskManagerGateway.setNotifyCheckpointCompleteConsumer(
+                (ignored0, ignored1, ignored2, ignored3) -> {
+                    throw new UnsupportedOperationException("notifyCheckpointCompleted was called");
+                });
+
+        final DefaultScheduler scheduler = createSchedulerAndStartScheduling(jobGraph);
+
+        final ExecutionAttemptID succeedingExecutionAttemptId =
+                Iterables.get(scheduler.getExecutionGraph().getAllExecutionVertices(), 0)
+                        .getCurrentExecutionAttempt()
+                        .getAttemptId();
+        final ExecutionAttemptID failingExecutionAttemptId =
+                Iterables.getLast(scheduler.getExecutionGraph().getAllExecutionVertices())
+                        .getCurrentExecutionAttempt()
+                        .getAttemptId();
+
+        // we have to make sure that the tasks are running before stop-with-savepoint is triggered
+        scheduler.updateTaskExecutionState(
+                new TaskExecutionState(
+                        jobGraph.getJobID(), failingExecutionAttemptId, ExecutionState.RUNNING));
+        scheduler.updateTaskExecutionState(
+                new TaskExecutionState(
+                        jobGraph.getJobID(), succeedingExecutionAttemptId, ExecutionState.RUNNING));
+
+        final CompletableFuture<String> stopWithSavepointFuture =
+                scheduler.stopWithSavepoint("savepoint-path", false);
+        checkpointTriggeredLatch.await();
+
+        final CheckpointCoordinator checkpointCoordinator = getCheckpointCoordinator(scheduler);
+
+        final AcknowledgeCheckpoint acknowledgeCheckpoint =
+                new AcknowledgeCheckpoint(jobGraph.getJobID(), succeedingExecutionAttemptId, 1);
+        final DeclineCheckpoint declineCheckpoint =
+                new DeclineCheckpoint(
+                        jobGraph.getJobID(),
+                        failingExecutionAttemptId,
+                        1,
+                        new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED));
+
+        checkpointCoordinator.receiveAcknowledgeMessage(acknowledgeCheckpoint, "unknown location");
+        checkpointCoordinator.receiveDeclineMessage(declineCheckpoint, "unknown location");
+
+        // we need to wait for the confirmations to be collected since this running in a separate
+        // thread
+        checkpointAbortionConfirmedLatch.await();
+
+        assertThat(
+                "Both of the executions where notified about the aborted checkpoint.",
+                executionAttemptIdsWithAbortedCheckpoint,
+                is(Arrays.asList(succeedingExecutionAttemptId, failingExecutionAttemptId)));
+
+        scheduler.updateTaskExecutionState(
+                new TaskExecutionState(
+                        jobGraph.getJobID(),
+                        succeedingExecutionAttemptId,
+                        ExecutionState.FINISHED));
```

Comment:
That is a mistake on my end - it proves that I didn't fully anticipate how the `DefaultSchedulerTest` needs to be implemented. Please correct me if I'm wrong: The tasks need to be cancelled, instead, since the cancellation call is send by the `TaskManager`. That will trigger the restart of the tasks which was initiated by the `CheckpointCoordinator`.

I came up with this code version (where the state changes to `FINISHED`/`FAILED`) due to the initial use-case of FLINK-21030. `FINISHED`/`FAILED` made the test to succeed as well because both are terminal states like `CANCEL`. Any terminal state triggers the restart of the tasks because the restart precedes the completion of the related `Execution`'s [releaseFuture](https://github.com/XComp/flink/blob/7ea786ac6af7286eb147449729316418b38cdafe/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java#L378).

I fixed it by transitioning to `CANCEL` state now.

Reply:
Only tasks which get canceled will need to return a `CANCELED` state. If this task does not need to be restarted, then it can simply continue running.

Reply:
Thanks for clarification.